working/2364 - primary constructors/feature-specification.md

```diff
@@ -654,30 +665,25 @@ contains a `<primaryConstructorNoConst>`, and the body of _D_ contains a
 `<constructorSignature>` beginning with `this` that contains a
 `<declaringParameterList>`.
 
-*That is, it is an error to have a declaring parameter list of a primary
-constructor both in the header and in the body.*
+*It is an error to have a declaring parameter list both in the header and
+in the body.*
 
 A compile-time error occurs if _D_ includes a `<classNamePart>` that
 does not contain a `<primaryConstructorNoConst>`, and the body of _D_
 contains a `<constructorSignature>` beginning with `this` that does not
 contain a `<declaringParameterList>`.
 
 *It is an error to have a declaring constructor in the class body, but
-no declaring parameter list, neither in the header nor in the body. Note
-that constant constructors are included because a
-`<constantConstructorSignature>` contains a `<constructorSignature>`.*
-
-A compile-time error occurs if _D_ includes a `<classNamePart>` beginning
-with `const`, and the body of _D_ contains a `<constructorSignature>`
-beginning with `this` which is not part of a
-`<constantConstructorSignature>`, and vice versa.
-
-*That is, it is an error for the header to contain `const` if there is a
-declaring constructor in the body as well, and it does not contain `const`;
-and it is an error for the header to not contain `const` if there is a
-declaring constructor in the body as well, and it does contain `const`.
-They must agree. Finally, it is of course allowed to omit `const` in both
-locations. In this case the constructor is not constant.*
+no declaring parameter list, neither in the header nor in the body.*
+
+A compile-time error occurs if a `<constantConstructorSignature>` starts
+with `const this` and no `<declaringParameterList>` is present.
```

Comment:
I'm not usually one to argue for encoding more syntax rules in the grammar, but this seems like a fairly easy one to handle, right? I think it's:

```
<constantConstructorSignature> ::= // Modified rule.
     'const' <constructorName> <formalParameterList>
   | 'const' 'this' ('.' <identifierOrNew>) <declaringParameterList>;
```